contracts/core/MatchExecutor.sol

```diff
@@ -0,0 +1,369 @@
+    }
+
+    if (externalFulfillments.nftsToTransfer.length > 0) {
+      /// Transfer the nfts to the intermediary
+      _transferMultipleNFTs(address(this), intermediary, externalFulfillments.nftsToTransfer);
+    }
+  }
+
+  /**
+   * @notice broker a trade by fulfilling orders on other exchanges and transferring nfts to the intermediary
+   * @param externalFulfillments The specification of the external calls to make and nfts to transfer
+   */
+  function _brokerCalldata(MatchExecutorTypes.ExternalFulfillments calldata externalFulfillments) internal {
+    uint256 numCalls = externalFulfillments.calls.length;
+    if (numCalls > 0) {
+      for (uint256 i; i < numCalls; ) {
+        _call(externalFulfillments.calls[i]);
+        unchecked {
+          ++i;
+        }
+      }
+    }
+
+    if (externalFulfillments.nftsToTransfer.length > 0) {
+      /// Transfer the nfts to the intermediary
+      _transferMultipleNFTs(address(this), intermediary, externalFulfillments.nftsToTransfer);
+    }
+  }
+
+  /**
+   * @notice Execute a call to the specified contract
+   * @param params The call to execute
+   */
+  function _call(MatchExecutorTypes.Call memory params) internal returns (bytes memory) {
+    if (params.isPayable) {
+      require(payableContracts[params.to], 'contract is not payable');
+      (bool _success, bytes memory _result) = params.to.call{value: params.value}(params.data);
+      require(_success);
+      return _result;
+    } else {
+      require(params.value == 0, 'value must be zero in a non-payable call');
+      (bool _success, bytes memory _result) = params.to.call(params.data);
+      require(_success);
+      return _result;
+    }
+  }
+
+  function _updateIntermediary(address _intermediary) internal {
+    require(_intermediary != address(0), 'intermdiary cannot be 0');
+    intermediary = _intermediary;
+    emit IntermediaryUpdated(_intermediary);
+  }
+
+  /**
+   * @notice Transfers multiple NFTs
+   * @param from the from address
+   * @param to the to address
+   * @param nfts nfts to transfer
+   */
+  function _transferMultipleNFTs(
+    address from,
+    address to,
+    OrderTypes.OrderItem[] memory nfts
+  ) internal {
+    for (uint256 i; i < nfts.length; ) {
+      _transferNFTs(from, to, nfts[i]);
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  /**
+   * @notice Transfer NFTs
+   * @dev Only supports ERC721, no ERC1155 or NFTs that conform to both ERC721 and ERC1155
+   * @param from address of the sender
+   * @param to address of the recipient
+   * @param item item to transfer
+   */
+  function _transferNFTs(
+    address from,
+    address to,
+    OrderTypes.OrderItem memory item
+  ) internal {
+    require(
+      IERC165(item.collection).supportsInterface(0x80ac58cd) && !IERC165(item.collection).supportsInterface(0xd9b67a26),
+      'only erc721'
+    );
+    _transferERC721s(from, to, item);
+  }
+
+  /**
+   * @notice Transfer ERC721s
+   * @dev requires approvals to be set
```

Comment:
who sets this approval? is this additional to approval already set on the main exchange?

Reply:
looks like this contract would be the owner of NFTs it is transferring, so no approval is needed

Reply:
Ah yes, removed this comment